### PR TITLE
Feature best match finder

### PR DESCRIPTION
**Pull request by gitwikc**
### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [ ] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### Best Match Finder Algorithm

Implemented the best match finder algorithm as discussed in the MITOCW [video lecture](https://youtu.be/5RSMLgy06Ew?list=PL6MpDZWD2gTF3mz26HSufmsIO-COKKb5j&t=1623) for the course 6.042 by Prof. Tom Leighton. The link above leads to the timestamp where he starts explaining how the algorithm works, using an example.

The problem is as follows:
- We are given two groups with same number of elements
- Each element has a *preference order*, ranking the elements in the descending order in which the element wants to pair with them, *i.e. 1st element is most preferred and last element is least preferred*
- We are given the task to find out such a matching of the elements such that there are no rogue couples
- *Rogue couple* is a set of two elements from different groups that have a higher preference for each other than the element they have currently been paired up with.
> **EXAMPLE**
> Let `g`, `f` be two sets whose elements are to be matched
> Let `(g1, f1)` and `(g2, f2)` be matched but 
> - `g2` has higher preference for `f1` than `f2` 
> - `f1` has higher preference for `g2` than `g1`
> In such a case, `g2` and `f1` are a **rogue couple**

The best matching is achieved when there are no rogue couples. Also for any two groups with the specifications as given above, there exists a best match. **Proof in video lecture**

### CODE REVIEW

File: other/best_match_finder.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    :param group2: Preferences of elements of group2 as a dict of the form
+    {<group2_elemt>: [...<List of group1_elemts in descending order of preference>]}
+    """
+    group1_checklist: dict[object, list[object]] = dict(
+        zip(group1.keys(), list(group1.values()).copy())
+    )
+    group2_possible_matches: dict[object, list[object]] = dict(
+        zip(group2.keys(), [[]] * 5)
+    )
+
```
Comment (CaedenPH) [inline other/best_match_finder.py:28]: You should typehint the return type of this function

**Review by CaedenPH — CHANGES_REQUESTED** [review]
You should typehint the function returns and potentially move the functions outside of the function and pass arguments. May make code more understandable

File: other/best_match_finder.py (hunk 1 above)
Comment (gitwikc) [reply other/best_match_finder.py:28]: Could you help me on how to typehint this one? This one is actually an assignment.

**Comment by gitwikc** [comment]
Added type hints, also created the test file and took out the nested functions

File: other/best_match_finder.py (hunk 1 above)
Comment (gitwikc) [reply other/best_match_finder.py]: Yes, I added the typehints for all functions now and took the nested functions out of the `get_best_match` function for more readability

**Comment by gitwikc** [comment]
I have written the code by import from the `typing` module in python

```
from typing import List, Dict, Tuple
```

But this bot keeps changing all the occurences of `List, Dict, Tuple` to `list, dict, tuple` respectively resulting in an error for unused imports. And using `list, dict, tuple` also gives an error for `types cannot be indexed`. What should I do now?

**Review by rohan472000 — CHANGES_REQUESTED** [review]
There are a few issues with the code:

1. The imports `typing.Dict`, `typing.List`, and `typing.Tuple` are unused and can be removed.
2. In `test_best_match_finder.py`, the line `if len(best_match) == True:` should be changed to `if len(best_match):` to follow PEP 8 guidelines.
3. In `test_best_match_finder.py`, `Dict` and `List` are undefined names. You should import them from `typing` or remove them altogether, depending on whether or not they are being used.
4. In `best_match_finder.py`, `zip` is being passed to the `list` function, which is not compatible. You can change it to `list(matches)` instead.

**Comment by tianyizheng02** [comment]
Closing due to lack of response to code review

**Pull request closed without merging**